packages/playwright/src/runner/watchMode.ts

```diff
@@ -15,130 +15,124 @@
   }
 
-  async onDirtyTestFiles(): Promise<void> {
-    if (this._dirtyTestFiles.size)
-      return;
-    await new Promise<void>(f => this._notifyDirtyFiles = f);
+  onopen(listener: () => void): void {
+    this.on('open', listener);
   }
 
-  takeDirtyTestFiles(): Map<FullProjectInternal, Set<string>> {
-    const result = this._dirtyTestFiles;
-    this._dirtyTestFiles = new Map();
-    return result;
+  send(data: string): void {
+    this._send(data);
   }
 }
 
-export async function runWatchModeLoop(config: FullConfigInternal): Promise<FullResult['status']> {
-  // Reset the settings that don't apply to watch.
-  config.cliPassWithNoTests = true;
-  for (const p of config.projects)
-    p.project.retries = 0;
-
-  // Perform global setup.
-  const testRun = new TestRun(config);
-  const taskRunner = createTaskRunnerForWatchSetup(config, [new ListReporter()]);
-  taskRunner.reporter.onConfigure(config.config);
-  const { status, cleanup: globalCleanup } = await taskRunner.runDeferCleanup(testRun, 0);
-  if (status !== 'passed')
-    await globalCleanup();
-  await taskRunner.reporter.onEnd({ status });
-  await taskRunner.reporter.onExit();
-  if (status !== 'passed')
-    return status;
-
-  // Prepare projects that will be watched, set up watcher.
-  const failedTestIdCollector = new Set<string>();
-  const originalWorkers = config.config.workers;
-  const fsWatcher = new FSWatcher();
-  await fsWatcher.update(config);
-
-  let lastRun: { type: 'changed' | 'regular' | 'failed', failedTestIds?: Set<string>, dirtyTestFiles?: Map<FullProjectInternal, Set<string>> } = { type: 'regular' };
+interface WatchModeOptions {
+  files?: string[];
+  projects?: string[];
+  grep?: string;
+  onlyChanged?: string;
+}
+
+export async function runWatchModeLoop(configLocation: ConfigLocation, initialOptions: WatchModeOptions): Promise<FullResult['status']> {
+  const options: WatchModeOptions = { ...initialOptions };
+
+  const testServerDispatcher = new TestServerDispatcher(configLocation);
+  const transport = new InMemoryTransport(
+      async data => {
+        const { id, method, params } = JSON.parse(data);
+        try {
+          const result = await testServerDispatcher.transport.dispatch(method, params);
+          transport.emit('message', JSON.stringify({ id, result }));
+        } catch (e) {
+          transport.emit('message', JSON.stringify({ id, error: String(e) }));
+        }
+      }
+  );
+  testServerDispatcher.transport.sendEvent = (method, params) => {
+    transport.emit('message', JSON.stringify({ method, params }));
+  };
+  const testServerConnection = new TestServerConnection(transport);
+  transport.emit('open');
+
+  const telesuiteUpdater = new TeleSuiteUpdater({ pathSeparator: path.sep, onUpdate() { } });
+
+  const dirtyTestFiles = new Set<string>();
+  const onDirtyTestFiles: { resolve?(): void } = {};
+
+  testServerConnection.onTestFilesChanged(async ({ testFiles: changedFiles }) => {
```

Comment:
I think we want to queue these operations, because listing tests might take longer than subsequent `onTestFilesChanged` events.